### ISSUE TITLE
Generic Request Pre-signing

### ISSUE THREAD

**Issue by rittneje** [aws/aws-sdk-net#1905]
## The Question

We need to create a pre-signed URL for STS GetCallerIdentity. We are able to do so easily with both the Golang SDK and Python (botocore). However, in the .NET SDK, we can only find pre-signing functionality specific to S3. Is generic pre-signing functionality missing?

---

This is a :question: general question

**Comment by ashishdhingra** [comment]
Hi @rittneje,

Good morning.

Thanks for posting guidance question. Yes, you are correct. Looking at https://github.com/aws/aws-sdk-net/tree/master/sdk/src/Services/SecurityToken, looks like create pre-signed URL for STS GetCallerIdentity is missing in .NET SDK. The Ruby `Presigner` version uses the custom implementation, refer https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-sts/presigner.rb). As a workaround, you may take some guidance from Ruby version and implement your helper method to generate presigned URL.

Hope this helps.

Thanks,
Ashish

**Comment by rittneje** [comment]
Does the .NET SDK expose Sigv4 itself? The Ruby code you linked references `Aws::Sigv4::Signer`. The Java, Python, and Golang SDKs also allow public access to it. I would prefer not to have to implement that manually.

**Comment by ashishdhingra** [comment]
> Does the .NET SDK expose Sigv4 itself? The Ruby code you linked references `Aws::Sigv4::Signer`. The Java, Python, and Golang SDKs also allow public access to it. I would prefer not to have to implement that manually.

@rittneje Please refer https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs for SigV4 signing.

**Comment by rittneje** [comment]
Thanks. Are we allowed to import that given that it is an "Internal" namespace? Or should we copy-paste the code?

**Comment by ashishdhingra** [comment]
> Thanks. Are we allowed to import that given that it is an "Internal" namespace? Or should we copy-paste the code?

@rittneje The SDK code is organized into various namespaces, `Amazon.Runtime.Internal.Auth` being one of them. The [AWS4Signer](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs) is used to sign all the requests before making service API call. You may import the namespace and use `AWS4Signer` class.

**Comment by rittneje** [comment]
Okay. That namespace does not show up in the docs though, so I wasn't sure. https://docs.aws.amazon.com/sdkfornet/v3/apidocs/index.html

**Comment by ashishdhingra** [comment]
> https://docs.aws.amazon.com/sdkfornet/v3/apidocs/index.html

@rittneje Pre-signing requests is an internal SDK operation and users do not have to pre-sign requests. Hence, [AWS4Signer](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs) might not have been specified in the API reference. Closing this issue for now since the necessary guidance has been provided.

**Comment by rittneje** [comment]
ping @ashishdhingra

**Comment by ashishdhingra** [comment]
> @ashishdhingra There are several standard tools such as Kubernetes/EKS that use a pre-signed GetCallerIdentity URL for authentication, so even just adding a new method to `AmazonSecurityTokenServiceClient` that takes an `IDictionary<string,string>` of additional headers and returns the pre-signed URL would be great.
> 
> Is this something the SDK team would be willing to accept via a PR?

Hi @rittneje,

Good morning.

The issue had been converted as feature request based on discussion with the team. While we cannot commit on the timeline by which this feature request would be prioritized for implementation given other items in the backlog as well, feel free to submit a PR which could be reviewed by the SDK team.

Thanks,
Ashish

**Comment by rittneje** [comment]
Don't close.

**Comment by tzachiabo** [comment]
@rittneje  are you planning to issue a PR to add this capability?

**Comment by rittneje** [comment]
@tzachiabo We might open one at least for GetCallerIdentity.

**Comment by GarrettBeatty** [comment]
Hi @rittneje,

  Good news — generic SigV4 signing and request pre-signing are now available as an API in `AWSSDK.Core` (namespace `Amazon.Runtime.Signing`), shipped in #4482. You no longer need to reach into the internal `AWS4Signer`. There are two
  entry points on `AWSSigV4Signer`:

  - `SignAsync(...)` → returns the headers (`Authorization`, `X-Amz-Date`, `X-Amz-Content-Sha256`, and `X-Amz-Security-Token` for temporary credentials) to attach to an outbound request.
  - `PresignAsync(..., expiry)` → returns a presigned `Uri` carrying all auth in the query string.

 **Presigned STS `GetCallerIdentity`**

  ```csharp
  using Amazon;
  using Amazon.Runtime.Credentials;
  using Amazon.Runtime.Signing;

  var region = RegionEndpoint.USEast1;

  var signingRequest = new AWSSigningRequest
  {
      HttpMethod = HttpMethod.Get,
      RequestUri = new Uri($"https://sts.{region.SystemName}.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15"),
  };

  var parameters = new AWSSigV4Parameters
  {
      Credentials = DefaultAWSCredentialsIdentityResolver.GetCredentials(),
      Region = region,
      Service = "sts",
  };

  PresignResult result = await AWSSigV4Signer.PresignAsync(signingRequest, parameters, TimeSpan.FromSeconds(60));

  // result.Uri is the presigned URL; fetch it with no extra headers.
  using var response = await new HttpClient().GetAsync(result.Uri);
  ```

**Comment by rittneje** [comment]
@GarrettBeatty I don't this entirely addressed the ask. I should not have to "hand-marshal" GetCallerIdentity like that, but the only other option is to reference the internal `GetCallerIdentityRequestMarshaller` class.
